### PR TITLE
Upgrade Registrar and its Beacon from Gnosis Safe

### PR DESCRIPTION
**Pull request by deep-quality-dev**

### CODE REVIEW

File: scripts/upgrade-from-safe.ts [hunk 1]
```diff
@@ -0,0 +1,97 @@
...
+  let beaconAddress: string;
+  if (hre.network.name === "goerli" || hre.network.name === "mainnet") {
+    const addresses = zer0ProtocolAddresses[hre.network.name];
+    if (!addresses) {
+      throw Error(`Network ${hre.network.name} not supported.`);
+    }
+    registrarAddress = addresses.zNS.registrar!;
+    beaconAddress = addresses.zNS.subregistrarBeacon!;
+    logger.log(`Registrar Proxy is at ${registrarAddress}`);
+    logger.log(`UpgradeableBeacon is at ${registrarAddress}`);
```
Comment (JamesEarle) [inline scripts/upgrade-from-safe.ts:36]: You are logging the `registrarAddress` twice here, please fix

File: scripts/upgrade-from-safe.ts [hunk 2]
```diff
@@ -0,0 +1,97 @@
...
+} from "@openzeppelin/hardhat-upgrades/dist/utils";
+import { getAdminAddress } from "@openzeppelin/upgrades-core";
+import { EthereumNetwork } from "./shared/config";
+import { proposeTransaction, verifyContract } from "./shared/helpers";
+
+const logger = getLogger("scripts::upgrade-from-safe");
+
+const GnosisSafeAddress: { [key in EthereumNetwork]: string } = {
+  goerli: "0x44B735109ECF3F1A5FE56F50b9874cEf5Ae52fEa",
+  mainnet: "", // todo
```
Comment (JamesEarle) [inline scripts/upgrade-from-safe.ts:18]: add the safe address here

0x1A1d3644fc9906B1EE3d35842789A83D33e99943

File: package.json [hunk 3]
```diff
@@ -2,6 +2,10 @@
   "devDependencies": {
     "@defi-wonderland/smock": "2.3.0",
     "@nomicfoundation/hardhat-network-helpers": "^1.0.7",
+    "@gnosis.pm/safe-core-sdk": "3.2.0",
```
Comment (JamesEarle) [inline package.json:5]: Looks like this package is deperecated. Consider migrating to `@safe-global/safe-core-sdk`

File: scripts/shared/helpers.ts [hunk 4]
```diff
@@ -38,3 +44,51 @@ export const calculateGasMargin = (
...
+    txServiceUrl: config.gnosisSafeServiceUrl[network],
+    ethAdapter,
+  });
+
+  const data = moduleInterface.encodeFunctionData(methodName, params);
+  const allTransactions = await safeService.getMultisigTransactions(
+    safe.getAddress()
+  );
+  const num =
+    allTransactions.count === 0 ? 1 : allTransactions.results[0].nonce + 1;
```
Comment (JamesEarle) [inline scripts/shared/helpers.ts:69]: is tx count for the gnosis safe 0-based? Why are we doing + 1 here?

**Review by JamesEarle — APPROVED** [review]
LGTM but note the package you're using is deprecated and the code should be migrated

https://www.npmjs.com/package/@gnosis.pm/safe-core-sdk

**Comment by JamesEarle** [comment]
Also, please be sure to resolve merge conflicts before you merge

File: scripts/shared/helpers.ts (hunk 4 above)
Comment (deep-quality-dev) [reply scripts/shared/helpers.ts:69]: Nonce is optional, I will remove it. If we need to pass nonce, we can get through calling `safeService.getNextNonce(gnosisSafe)`

**Pull request closed without merging**